### PR TITLE
PERF-XDP-TX-FORWARD-MULTICORE/SINGLECORE: New testcases to check XDP Forwarding performance.

### PR DESCRIPTION
**Pull request by abhimarathe**
This test starts pktgen and checks XDP_TX forwarding performance by starting xdpdump application in forwarding configuration on forwarder VM and checks how many packets received at the receiver interface by running xdpdump application in drop configuration (number of packets received == number of packets dropped).
```
[LISAv2 Test Results Summary]
Test Run On           : 08/26/2020 23:34:14
ARM Image Under Test  : Canonical : UbuntuServer : 18.04-LTS : latest
Total Test Cases      : 1 (1 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:7


   ID TestArea             TestCaseName                                                                TestResult TestDuration(in minutes)
-------------------------------------------------------------------------------------------------------------------------------------------
    1 NETWORK              PERF-XDP-TX-FORWARD-MULTICORE                                                     PASS                 5.53
      ARMImageName: Canonical UbuntuServer 18.04-LTS latest, Networking: SRIOV, OverrideVMSize: Standard_D64_v3, TestLocation: eastus2, Kernel Version: 5.7.0


Logs can be found at C:\LISAv2\YF43\TestResults\2020-08-26-16-33-27-1417
[LISAv2 Test Results Summary]
Test Run On           : 08/27/2020 06:56:44
ARM Image Under Test  : Canonical : UbuntuServer : 18.04-LTS : latest
Total Test Cases      : 1 (1 Passed, 0 Failed, 0 Aborted, 0 Skipped)

   ID TestArea             TestCaseName                                                                TestResult TestDuration(in minutes)
-------------------------------------------------------------------------------------------------------------------------------------------
    1 NETWORK              PERF-XDP-TX-FORWARD-SINGLECORE                                                    PASS                  5.3
      ARMImageName: Canonical UbuntuServer 18.04-LTS latest, Networking: SRIOV, OverrideVMSize: Standard_D64_v3, TestLocation: eastus2, Kernel Version: 5.7.0


Logs can be found at C:\LISAv2\QY33\TestResults\2020-08-26-23-55-53-2247
```

### CODE REVIEW

**Comment by abhimarathe** [comment]
```
[LISAv2 Test Results Summary]
Test Run On           : 08/27/2020 19:48:16
ARM Image Under Test  : Canonical : UbuntuServer : 18.04-LTS : latest
Total Test Cases      : 1 (1 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:30

   ID TestArea             TestCaseName                                                                TestResult TestDuration(in minutes)
-------------------------------------------------------------------------------------------------------------------------------------------
    1 NETWORK              PERF-XDP-TX-FORWARD-SINGLECORE                                                    PASS                10.72
      ARMImageName: Canonical UbuntuServer 18.04-LTS latest, Networking: SRIOV, OverrideVMSize: Standard_D64_v3, TestLocation: eastus2, Kernel Version: 5.3.0-1035-azure -> 5.7.0


Logs can be found at C:\LISAv2\RQ77\TestResults\2020-08-27-12-47-07-2007
[LISAv2 Test Results Summary]
Test Run On           : 08/27/2020 15:47:00
ARM Image Under Test  : Canonical : UbuntuServer : 18.04-LTS : latest
Total Test Cases      : 1 (1 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:24

   ID TestArea             TestCaseName                                                                TestResult TestDuration(in minutes) 
-------------------------------------------------------------------------------------------------------------------------------------------
    1 NETWORK              PERF-XDP-TCP-RX-DROP-SINGLECORE                                                   PASS                 6.67 
      ARMImageName: Canonical UbuntuServer 18.04-LTS latest, Networking: SRIOV, OverrideVMSize: Standard_D64_v3, TestLocation: eastus2, Kernel Version: 5.3.0-1035-azure -> 5.7.0


Logs can be found at C:\LISAv2\TN20\TestResults\2020-08-27-08-46-15-9484
```

**Comment by abhimarathe** [comment]
```
[LISAv2 Test Results Summary]
Test Run On           : 08/28/2020 18:58:51
ARM Image Under Test  : Canonical : 0001-com-ubuntu-server-focal : 20_04-lts : latest
Total Test Cases      : 1 (1 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:15

   ID TestArea             TestCaseName                                                                TestResult TestDuration(in minutes) 
-------------------------------------------------------------------------------------------------------------------------------------------
    1 NETWORK              PERF-XDP-TX-FORWARD-SINGLECORE                                                    PASS                 8.85 
      ARMImageName: Canonical 0001-com-ubuntu-server-focal 20_04-lts latest, Networking: SRIOV, OverrideVMSize: Standard_D64_v3, TestLocation: eastus2, Kernel Version: 5.4.0-1022-azure


Logs can be found at C:\LISAv2\EG27\TestResults\2020-08-28-11-57-03-0461
```

File: Testscripts/Linux/XDPForwardingTest.sh [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+                        MACarr="$MACarr, $(printf '0x%s\n' ${ADDR[$i]})";
+                done
+        done <<< "$1"
+        echo "$MACarr"
+}
+
+function configure_XDPDUMP_TX(){
+        LogMsg "Configuring TX Setup"
+        # new distros does not have ifconfig present by default
+	LogMsg "Installing net-tools for confirming ifconfig is present in VM."
```
Comment (juhlee-microsoft) [inline Testscripts/Linux/XDPForwardingTest.sh]: indentation? Some lines are 8-char size and others 4-char size.

Comment (juhlee-microsoft) [reply Testscripts/Linux/XDPForwardingTest.sh]: image

File: Testscripts/Linux/XDPForwardingTest.sh [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+# Source utils.sh
+. ${UTIL_FILE} || {
+    echo "ERROR: unable to source ${UTIL_FILE}!"
+    echo "TestAborted" > state.txt
+    exit 0
+}
+
+XDPUTIL_FILE="./XDPUtils.sh"
+
+# Source utils.sh
```
Comment (juhlee-microsoft) [inline Testscripts/Linux/XDPForwardingTest.sh]: XDPUtils.sh ?

**Review by juhlee-microsoft — CHANGES_REQUESTED** [review]

File: Testscripts/Linux/XDPForwardingTest.sh (hunk 1 above)
Comment (abhimarathe) [reply Testscripts/Linux/XDPForwardingTest.sh]: fixed.

File: Testscripts/Linux/XDPForwardingTest.sh (hunk 2 above)
Comment (abhimarathe) [reply Testscripts/Linux/XDPForwardingTest.sh]: fixed.

**Review by juhlee-microsoft — DISMISSED** [review]

File: Testscripts/Linux/XDPForwardingTest.sh [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+. ${XDPUTIL_FILE} || {
+    echo "ERROR: unable to source ${XDPUTIL_FILE}!"
+    echo "TestAborted" > state.txt
```
Comment (lubaihua33) [inline Testscripts/Linux/XDPForwardingTest.sh]: We have this function "SetTestStateAborted" in utils.sh.

File: Testscripts/Linux/XDPForwardingTest.sh [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+        LogMsg "Updated Source &  Destination IP address in $xdpdumpFileName on $forwarder"
+}
+
+UTIL_FILE="./utils.sh"
+
+# Source utils.sh
+. ${UTIL_FILE} || {
+    echo "ERROR: unable to source ${UTIL_FILE}!"
+    echo "TestAborted" > state.txt
+    exit 0
```
Comment (lubaihua33) [inline Testscripts/Linux/XDPForwardingTest.sh:72]: Please check the indentation, not the same with others.

File: Testscripts/Linux/XDPForwardingTest.sh (hunk 3 above)
Comment (lubaihua33) [reply Testscripts/Linux/XDPForwardingTest.sh]: And check the indentation.

File: Testscripts/Linux/XDPUtils.sh [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        vfForward=$(ethtool -S $vfName | grep rx_xdp_tx_xmit | cut -d':' -f2)
+        echo "$((vfForward))"
+}
+
+function download_pktgen_scripts(){
+        local ip=$1
+        local dir=$2
+        local cores=$3
+        if [ "${cores}" = "multi" ];then
+                ssh $ip "wget https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/plain/samples/pktgen/pktgen_sample05_flow_per_thread.sh?h=v5.7.8 -O ${dir}/pktgen_sample.sh"
```
Comment (lubaihua33) [inline Testscripts/Linux/XDPUtils.sh]: I see you used many ssh "ip" "command“ . We have a function "Run_SSHCommand" in utils.sh. You could check if this function can be used.

File: Testscripts/Linux/pktgenSetup.sh [hunk 6]
```diff
@@ -73,6 +14,15 @@ UTIL_FILE="./utils.sh"
         exit 0
 }
 
+XDPUTIL_FILE="./XDPUtils.sh"
+
+# Source utils.sh
```
Comment (lubaihua33) [inline Testscripts/Linux/pktgenSetup.sh]: Need to change this comment.

File: Testscripts/Linux/pktgenSetup.sh [hunk 7]
```diff
@@ -73,6 +14,15 @@ UTIL_FILE="./utils.sh"
...
+. ${XDPUTIL_FILE} || {
+    echo "ERROR: unable to source ${XDPUTIL_FILE}!"
+    echo "TestAborted" > state.txt
+    exit 0
```
Comment (lubaihua33) [inline Testscripts/Linux/pktgenSetup.sh:23]: Indentation.

**Review by lubaihua33 — CHANGES_REQUESTED** [review]

File: Testscripts/Linux/pktgenSetup.sh (hunk 7 above)
Comment (abhimarathe) [reply Testscripts/Linux/pktgenSetup.sh:23]: indentation is correct 4 spaces

File: Testscripts/Linux/pktgenSetup.sh (hunk 6 above)
Comment (abhimarathe) [reply Testscripts/Linux/pktgenSetup.sh]: fixed

File: Testscripts/Linux/XDPForwardingTest.sh (hunk 4 above)
Comment (abhimarathe) [reply Testscripts/Linux/XDPForwardingTest.sh:72]: Indentation is 4 spaces

File: Testscripts/Linux/XDPForwardingTest.sh (hunk 3 above)
Comment (abhimarathe) [reply Testscripts/Linux/XDPForwardingTest.sh]: Fixed

File: Testscripts/Linux/XDPForwardingTest.sh [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+forwarderSecondMAC=$((ssh ${forwarder} "ip link show ${nicName}") | grep ether | awk '{print $2}')
+LogMsg "Forwarder second MAC: ${forwarderSecondMAC}"
+if [ "${core}" = "single" ];then
+    startCommand="cd ${pktgenDir} && ./pktgen_sample.sh -i ${nicName} -m ${forwarderSecondMAC} -d ${forwarderSecondIP} -v -n${packetCount}"
+    LogMsg "Starting pktgen on sender: $startCommand"
+    ssh ${sender} "modprobe pktgen; lsmod | grep pktgen"
+    result=$(ssh ${sender} "${startCommand}")
+else
+    startCommand="cd ${pktgenDir} && ./pktgen_sample.sh -i ${nicName} -m ${forwarderSecondMAC} -d ${forwarderSecondIP} -v -n${packetCount} -t8"
+    LogMsg "Starting pktgen on sender: ${startCommand}"
```
Comment (harz566) [inline Testscripts/Linux/XDPForwardingTest.sh]: dup code with above? it seems just different startcommand should be ok, all others following steps are the same in this simple if else fi.

File: Testscripts/Linux/XDPForwardingTest.sh [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+check_exit_status "Build xdpdump with DROP Action on ${receiver}"
+
+# Calculate packet drops before tests
+packetDropBefore=$(ssh ${receiver} ". XDPUtils.sh && calculate_packets_drop ${nicName}")
+LogMsg "Before test, Packet drop count on ${receiver} is ${packetDropBefore}"
+# Calculate packets forwarded before tests
+pktForwardBefore=$(ssh ${forwarder} ". XDPUtils.sh && calculate_packets_forward ${nicName}")
+LogMsg "Before test, Packet forward count on ${forwarder} is ${pktForwardBefore}"
+
+# Start XDPDump on receiver
```
Comment (harz566) [inline Testscripts/Linux/XDPForwardingTest.sh:122]: like this: 'Start XDPDump on receiver',  'Start XDPDump on forwarder' ; a few other places have similar 'dup' code, not exact the same, but 90% codes are same, could we wrap with some shared function to minimize maintenance cost?

File: Testscripts/Linux/pktgenSetup.sh [hunk 10]
```diff
@@ -118,36 +70,36 @@ ssh -f ${client} "sh -c '${xdpdumpCommand} &'"
...
 LogMsg "Before Drop: $pakcetDropBefore After Drop:$pakcetDropAfter "
 packetsDropped=$((pakcetDropAfter - pakcetDropBefore))
 LogMsg "Pakcets dropped: $packetsDropped"
 dropLimit=$(( packetCount*packetDropThreshold/100 ))
 if [ $packetsDropped -lt $dropLimit ]; then
-        LogErr "receiver did not receive packets."
-        SetTestStateAborted
-        exit 1
+    LogErr "receiver did not receive packets."
+    SetTestStateAborted
```
Comment (harz566) [inline Testscripts/Linux/pktgenSetup.sh]: Failed?

File: Testscripts/Windows/PERF-XDP-TX-FORWARD.ps1 [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+        # TODO: Check max installation time for other distros when added
+        $timer = 0
+        while ($testJob -and ((Get-Job -Id $testJob).State -eq "Running")) {
+            $currentStatus = Run-LinuxCmd -ip $masterVM.PublicIP -port $masterVM.SSHPort `
+                -username $user -password $password -command "tail -2 ~/xdpConsoleLogs.txt | head -1" -runAsSudo
+            Write-LogInfo "Current Test Status: $currentStatus"
+            Wait-Time -seconds 20
+            $timer += 1
+            if ($timer -gt 30) {
+                Throw "XDPFowardingTest did not stop after 10 mins. Please check xdpConsoleLogs."
```
Comment (harz566) [inline Testscripts/Windows/PERF-XDP-TX-FORWARD.ps1:76]: if we want to use limit '10' minutes, we should use stopwatch, actually, using current logic in the powershell, it's definitely over 10 minutes.  But just a suggestion only, not a big issue, we can still leave it as it is, if you think it doesn't matter or just no time to change it.

File: Testscripts/Windows/PERF-XDP-TX-FORWARD.ps1 [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+
+        if ($currentState -imatch "TestCompleted") {
+            Write-LogInfo "Test Completed"
+            $testResult = "PASS"
+        }
+        elseif ($currentState -imatch "TestAborted") {
+            Write-LogErr "Test Aborted. Last known status: $currentStatus."
+            $testResult = "ABORTED"
+        }
+        elseif ($currentState -imatch "TestSkipped") {
```
Comment (harz566) [inline Testscripts/Windows/PERF-XDP-TX-FORWARD.ps1:91]: seems we have a common method to get the translated result.   Function Get-FinalResultHeader()

Comment (abhimarathe) [reply Testscripts/Windows/PERF-XDP-TX-FORWARD.ps1:91]: It is a bit different. Does not display logs which I think is important to notify that the script has ran and Reason behind Success/Failure is Last known status. Please let me know your thoughts on this.

File: Testscripts/Windows/PERF-XDP-TX-FORWARD.ps1 (hunk 11 above)
Comment (abhimarathe) [reply Testscripts/Windows/PERF-XDP-TX-FORWARD.ps1:76]: It does not have any hard time limit, so I would like to leave it as it is. but thank you for  educating me on this. 😄

**Comment by abhimarathe** [comment]
```
[LISAv2 Test Results Summary]
Test Run On           : 08/29/2020 15:56:25
ARM Image Under Test  : Canonical : 0001-com-ubuntu-server-focal : 20_04-lts : latest
Total Test Cases      : 2 (2 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:26

   ID TestArea             TestCaseName                                                                TestResult TestDuration(in minutes)
-------------------------------------------------------------------------------------------------------------------------------------------
    1 NETWORK              PERF-XDP-TX-FORWARD-SINGLECORE                                                    PASS                 8.69
      ARMImageName: Canonical 0001-com-ubuntu-server-focal 20_04-lts latest, Networking: SRIOV, OverrideVMSize: Standard_D64_v3, TestLocation: eastus2, Kernel Version: 5.4.0-1022-azure
    2 NETWORK              PERF-XDP-RX-DROP-SINGLECORE                                                       PASS                 5.59
      ARMImageName: Canonical 0001-com-ubuntu-server-focal 20_04-lts latest, Networking: SRIOV, OverrideVMSize: Standard_D64_v3, TestLocation: eastus2, Kernel Version: 5.4.0-1022-azure


Logs can be found at C:\LISAv2\JK50\TestResults\2020-08-29-08-55-39-9406
```

File: Testscripts/Linux/XDPForwardingTest.sh (hunk 8 above)
Comment (abhimarathe) [reply Testscripts/Linux/XDPForwardingTest.sh]: fixed

File: Testscripts/Linux/XDPForwardingTest.sh (hunk 9 above)
Comment (abhimarathe) [reply Testscripts/Linux/XDPForwardingTest.sh:122]: fixed

**Review by lubaihua33 — DISMISSED** [review]

**Review by juhlee-microsoft — DISMISSED** [review]

**Comment by juhlee-microsoft** [comment]
Please fix the broken part in the PR Checker

**Comment by abhimarathe** [comment]
> Please fix the broken part in the PR Checker

PR Checker failed as I have used pktgenResult variable in XDPUtils.sh to store pktgen result. I'm using it in other script (XDPForwardingTest.sh and pktgenSetup.sh) but not in the same shell script which is why it is failing.
Any suggestions?
I know we can use echo to return value from a function but in that case I can't use LogMsg inside function and that's necessary as which pktgen command has ran should be captured in the logs.

**Review by harz566 — DISMISSED** [review]

**Review by lubaihua33 — DISMISSED** [review]

**Review by juhlee-microsoft — APPROVED** [review]

**Review by harz566 — APPROVED** [review]

**Review by lubaihua33 — APPROVED** [review]

**Pull request merged**